source/server/CMakeLists.txt

```diff
@@ -0,0 +1,25 @@
+add_library(ni_grpc_device_server_core
+  "device_enumerator.cpp"
+  "feature_toggles.cpp"
+  "logging.cpp"
+  "semaphore.cpp"
+  "server_configuration_parser.cpp"
+  "server_security_configuration.cpp"
+  "session_repository.cpp"
+  "session_utilities_service.cpp"
+  "shared_library.cpp"
+  "syscfg_library.cpp")
+
+target_include_directories(ni_grpc_device_server_core
+  PUBLIC ${CMAKE_SOURCE_DIR}/imports/include
+  PUBLIC "../" # HACK should be able to sanitize this, but include statements expect it to be server/<whatever>.h
```

Comment:
Note: these libraries don't *need* to be in separate CmakeLists.txt files. They could be moved back to the top-level CMakeLists.txt.

I think the split worked a little better for core_proto than the server. For this to really be coherent, we would probably need to move the source folders around some (?). Is an ooch change like this worth it? Or is it better to leave it in the top-level unless/until we spec out how we want to organize it?

Obviously any auto-generated driver-specific CMakeLists.txt files *would* need to be separate files.

Reply:
Also, any auto-generated CMakeLists.txt files would need to be generated at-or-before cmake configure time. Not during the build like the current codegen pass.

That’s something to work through, but I think it’s ok. The rest of the codegen can still run during the build. So youd only need to reconfigure if you’re adding a new generated output file or dependency. Right now, to do that, you need to (a) update the metadata, (b) update cmakelists, (c) re-run configure. 

So we wouldn’t be adding a step there and would make the process easier in the normal case (add a new service or opt into an existing feature) and more powerful in novel cases.

Reply:
We could have the codegen run as part of our workflow, or even add a script that did the generation and CMake configure instead. Having the generated files be created by CMake was a trick to make sure they were always up to date after metadata changes but it might not have been the best approach.

Reply:
Yeah, that's a good point. Running during build is definitely useful during development. And it semi-enforces that the generated source is up-to-date, which is also useful for development. I could imagine creating a separate "build.py" script that generates code and then invokes cmake. Especially if we started doing more python stuff (venvs, pytest, etc) so that it's more awkward for cmake to drive the process. And then we'd just have to retrain ourselves to use that instead of cmake (or to use a custom build action instead of the cmake vscode plugin). ... And figure out how to handle the different cmake configurations we use (cross-compile, ninja generator, -j, etc).

We've also been lucky (?) that codegen is still relatively cheap to run (cheaper than protoc, cheaper than compiling the generated code).

For generating cmakelists. It's pretty easy to add an `execute_process` to run during configure. I think it would also be worthwhile to re-run the codegen during build and "fail on change" or something like that.

Reply:
For the "fail on change" comment - I've considered something like that in the past. Specifically, I think our PR builds should fail if the codegen files change during the build (e.g., someone didn't check-in the changed code-gen files). It's usually easy to catch but that would be another nice to have in the future.